cmd/reconciler/main.go

```diff
@@ -131,6 +159,122 @@ func main() {
 		status.EnablePanicOnMisuse()
 	}
 
+	reconcilerScope := declared.Scope(*scope)
+	fight.SetFightThreshold(*fightDetectionThreshold)
+
+	cl, discoveryClient, baseApplier, supervisor, err := configureClients(reconcilerScope)
+	if err != nil {
+		klog.Error(err)
+	}
+
+	// Start listening to signals
+	signalCtx := signals.SetupSignalHandler()
+	mgrOptions := ctrl.Options{
+		Scheme: core.Scheme,
+		BaseContext: func() context.Context {
+			return signalCtx
+		},
+	}
+	// For Namespaced Reconcilers, set the default namespace to watch.
+	// Otherwise, all namespaced informers will watch at the cluster-scope.
+	// This prevents Namespaced Reconcilers from needing cluster-scoped read
+	// permissions.
+	if reconcilerScope != declared.RootReconciler {
+		mgrOptions.Namespace = string(reconcilerScope)
+	}
+
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), mgrOptions)
+	if err != nil {
+		klog.Fatalf("Error starting manager: %v", err)
+	}
+
+	// This cancelFunc will be used by the Finalizer to stop all the other
+	// controllers (Syncer & Remediator).
+	controllerCtx, stopControllers := context.WithCancel(signalCtx)
+
+	// These channels will be closed when the corresponding controllers have exited,
+	// signalling for the finalizer to continue.
+	doneChannelForSyncer := make(chan struct{})
+	doneChannelForRemediator := make(chan struct{})
+
+	// Create the Finalizer
+	// The caching client built by the controller-manager doesn't update
+	// the GET reconcilerState on UPDATE/PATCH. So we need to use the non-caching client
+	// for the finalizer, which does GET/LIST after UPDATE/PATCH.
+	f := finalizer.New(reconcilerScope, supervisor, cl, // non-caching client
+		stopControllers, doneChannelForSyncer, doneChannelForRemediator)
+
+	// Create the Finalizer Controller
+	finalizerController := &finalizer.Controller{
+		SyncScope: reconcilerScope,
+		SyncName:  *syncName,
+		Client:    mgr.GetClient(), // caching client
+		Scheme:    mgr.GetScheme(),
+		Mapper:    mgr.GetRESTMapper(),
+		Finalizer: f,
+	}
+
+	// Register the Finalizer Controller
+	if err = finalizerController.SetupWithManager(mgr); err != nil {
+		klog.Fatalf("Instantiating Finalizer: %v", err)
+	}
+
+	// Configure Syncer and Remediator.
+
+	// Get a separate config for the remediator to talk to the apiserver since
+	// we want a longer REST config timeout for the remediator to avoid restarting
+	// idle watches too frequently.
+	cfgForWatch, err := restconfig.NewRestConfig(watch.RESTConfigTimeout)
```

Comment:
Maybe not relevant to this PR, but I think we might need to send multiple clients to the remediator at some point to separate the watch timeout and list timeout from all the other operations. Or maybe we needs to be more explicit about those timeouts when we make those requests. Otherwise they all get the 1hr timeout.